### PR TITLE
Configure resources & auto-tune flags

### PR DESCRIPTION
**Pull request by fabxc**
@brancz

### CODE REVIEW

File: pkg/operator/petset.go [hunk 1]
```diff
@@ -208,6 +219,12 @@ func makePetSetSpec(name, image, version, retention string, replicas int32, aler
 								MountPath: "/etc/prometheus/config",
 							},
 						},
+						Resources: v1.ResourceRequirements{
```
Comment (fabxc) [inline pkg/operator/petset.go]: Ah right, will fix.
We can bump it to 10. I intentionally didn't apply a limit because what knows why it might burst.

Comment (brancz) [inline pkg/operator/petset.go]: I'll keep an eye on the mem usage, regarding cpuquota this should be fine though

File: pkg/operator/petset.go [hunk 2]
```diff
@@ -157,8 +165,10 @@ func makePetSetSpec(name, image, version, retention string, replicas int32, aler
 							},
 						},
 						Args: []string{
-							"-storage.local.retention=" + retention,
-							"-storage.local.memory-chunks=500000",
+							"-storage.local.retention=" + p.Spec.Retention,
+							"-storage.local.memory-chunks=" + fmt.Sprintf("%d", memChunks),
+							"-storage.local.max-chunks-to-persist" + fmt.Sprintf("%d", memChunks/2),
+							"-storage.local.num-fingerprint-mutexes=4096",
```
Comment (brancz) [inline pkg/operator/petset.go]: Why set this value to 4096 if that's already the default?

File: pkg/operator/petset.go [hunk 3]
```diff
@@ -126,19 +122,31 @@ func makePetSetService(p *spec.Prometheus) *v1.Service {
...
+	// requested memory can fit. The user has to specify an appropriate buffering
+	// in memory limits to catch increased memory usage during query bursts.
+	// More info: https://prometheus.io/docs/operating/storage/.
+	reqMem := p.Spec.Resources.Requests[v1.ResourceMemory]
+	// 1024 byte is the fixed chunk size. With increasing number of chunks actually
+	// in memory, overhead owed to their management, higher ingestion buffers, etc.
+	// increases.
+	// We are conservative for now an assume this to be 80% as the Kubernetes environment
+	// generally has a very high time series churn.
+	memChunks := reqMem.Value() / 1024 / 5
```
Comment (brancz) [inline pkg/operator/petset.go]: I likely don't understand these flags or memory behavior correctly, but with this calculation looks more like it is 20% of the requested memory in chunks, is that intentional?

Comment (fabxc) [inline pkg/operator/petset.go]: Yea, see prometheus/prometheus#455 for long-running tracking discussion.
Lot's of things not really understood about this as there are so many factors. Especially the time series churn in Kubernetes.

20% seems safe for now. 30% is the go-to number at the moment.
Obviously the data held in memory should be more at the order of 80% – hopefully we can address this upstream at some point.

File: pkg/operator/petset.go (hunk 2 above)
Comment (fabxc) [inline pkg/operator/petset.go]: Ah, I didn't even check the default. Unnecessary but being explicit about it doesn't hurt I suppose.

File: pkg/operator/petset.go (hunk 1 above)
Comment (fabxc) [inline pkg/operator/petset.go]: Fixed.

**Comment by fabxc** [comment]
Also setting "2Gi" for requested memory now if it's not specified. In the end we have to specify something for some Prometheus flags to be set or use the default, which should then also be somehow reflected in requested memory to the scheduler.

**Comment by fabxc** [comment]
Merging. LMK if you have further comments.

**Pull request merged**